### PR TITLE
[Deepspeed] Allow HF optimizer and scheduler to be passed to deepspeed 

### PR DESCRIPTION
**Pull request by cli99**
# Use HF optimizer and/or scheduler unless specified in deepspeed config

If HF is already creating an optimizer and LR scheduler, we should not try to match that config/implementation in a ds_config.json instead we pass it to deepspeed.initialize(..., lr_scheduler=hf_lr_scheduler)

* [x] This PR checks if ds_config has an optimizer or scheduler, if it does not, calls  `create_optimizer` or `create_cheduler` (after splitting it) to create an optimizer or scheduler.  Then HF optimizer and scheduler are passed it to deepspeed.initialize().
DeepSpeed can handle any optimizer and scheduler if these are passed directly to deepspeed.initialize() as an object.

Due to the chicken-n-egg init problem, the valid combinations are:

| Combos       | HF Scheduler | DS Scheduler |
|--------------|--------------|--------------|
| HF Optimizer | Yes          | Yes          |
| DS Optimizer | No           | Yes          |

but if `cpu_offload` is used all bets are off - we can only use DS optim/sched.

----------
added by @stas00 below:

Added:
* [x] make init_deepspeed support config dict, besides the config file - this makes the testing much easier
* [x] add tests for this PR using this new feature of passing the dict
* [x] various small clean ups
* [x] update the docs
* [x] check for `cpu_offload` - add test
* [x] recode the config overrides to have one true source of values
* [x] tweak one not working test 

**blocking event: waiting for a new release 0.3.13 from DeepSpeed.**

@sgugger

### CODE REVIEW

File: src/transformers/integrations.py
Comment (stas00) [inline src/transformers/integrations.py:383]: These need to be conditional - only passed if the config didn't have those already.

File: src/transformers/trainer.py
Comment (stas00) [inline src/transformers/trainer.py:669]: If we are making these public methods, this one  would need a docstring too.

File: src/transformers/trainer_utils.py
Comment (stas00) [inline src/transformers/trainer_utils.py:428]: Thank for catching those discrepancies.

Ideally let's try not to cross unrelated areas in the same PR, but I suppose it's OK for now.

**Review by stas00 — APPROVED** [review]
That's a very good proposal, @cli99.

So are you saying DeepSpeed will handle any optimizer and scheduler if these are passed directly as an object?

One more thing we need is a test, it probably would be easier for you if I added it since I need to rework a few things to support this test. But you will need to switch it to normal PR from draft for me to be able to push into it. We usually don't use Draft but edit the PR title to start with `[WIP]` work in progress.

File: src/transformers/integrations.py
Comment (cli99) [reply src/transformers/integrations.py]: Set it to None if the config has an optimizer. Same thing for lr_scheduler

File: src/transformers/trainer.py
Comment (cli99) [reply src/transformers/trainer.py:669]: Added docstring

File: src/transformers/integrations.py
Comment (stas00) [reply src/transformers/integrations.py]: That is a simpler solution!

**Comment by stas00** [comment]
OK, 2 tests added and no, this doesn't work w/o neither default optimizer nor scheduler.

```
pytest -sv examples/tests/deepspeed/test_deepspeed.py -k test_hf_native_scheduler

examples/tests/deepspeed/test_deepspeed.py:103:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
src/transformers/trainer.py:917: in train
    model, optimizer, lr_scheduler = init_deepspeed(self, num_training_steps=max_steps)
src/transformers/integrations.py:351: in init_deepspeed
    trainer.create_scheduler(num_training_steps=num_training_steps)
src/transformers/trainer.py:685: in create_scheduler
    self.lr_scheduler = get_scheduler(
src/transformers/optimization.py:266: in get_scheduler
    return schedule_func(optimizer, num_warmup_steps=num_warmup_steps, num_training_steps=num_training_steps)
src/transformers/optimization.py:98: in get_linear_schedule_with_warmup
    return LambdaLR(optimizer, lr_lambda, last_epoch)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <torch.optim.lr_scheduler.LambdaLR object at 0x7fd86fb0a160>, optimizer = None
lr_lambda = <function get_linear_schedule_with_warmup.<locals>.lr_lambda at 0x7fd86fafc160>, last_epoch = -1, verbose = False

    def __init__(self, optimizer, lr_lambda, last_epoch=-1, verbose=False):
        self.optimizer = optimizer

        if not isinstance(lr_lambda, list) and not isinstance(lr_lambda, tuple):
>           self.lr_lambdas = [lr_lambda] * len(optimizer.param_groups)
E           AttributeError: 'NoneType' object has no attribute 'param_groups'

/home/stas/anaconda3/envs/main-38/lib/python3.8/site-packages/torch/optim/lr_scheduler.py:197: AttributeError
````


```
pytest -sv examples/tests/deepspeed/test_deepspeed.py -k test_hf_native_optimizer


examples/tests/deepspeed/test_deepspeed.py:91: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
src/transformers/trainer.py:917: in train
    model, optimizer, lr_scheduler = init_deepspeed(self, num_training_steps=max_steps)
src/transformers/integrations.py:384: in init_deepspeed
    model, optimizer, _, lr_scheduler = deepspeed.initialize(
../../github/00optimize/DeepSpeed/deepspeed/__init__.py:110: in initialize
    engine = DeepSpeedEngine(args=args,
../../github/00optimize/DeepSpeed/deepspeed/runtime/engine.py:174: in __init__
    self._configure_optimizer(optimizer, model_parameters)
../../github/00optimize/DeepSpeed/deepspeed/runtime/engine.py:570: in _configure_optimizer
    self.optimizer = self._configure_zero_optimizer(basic_optimizer)
../../github/00optimize/DeepSpeed/deepspeed/runtime/engine.py:691: in _configure_zero_optimizer
    optimizer = FP16_DeepSpeedZeroOptimizer(
../../github/00optimize/DeepSpeed/deepspeed/runtime/zero/stage2.py:239: in __init__
    flatten_dense_tensors_aligned(
../../github/00optimize/DeepSpeed/deepspeed/runtime/zero/stage2.py:74: in flatten_dense_tensors_aligned
    return _flatten_dense_tensors(padded_tensor_list)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

tensors = []

    def _flatten_dense_tensors(tensors):
        """Flatten dense tensors into a contiguous 1D buffer. Assume tensors are of
        same dense type.
    
        Since inputs are dense, the resulting tensor will be a concatenated 1D
        buffer. Element-wise operation on this buffer will be equivalent to
        operating individually.
    
        Args:
            tensors (Iterable[Tensor]): dense tensors to flatten.

        Returns:
            A contiguous 1D buffer containing input tensors.
        """
        if len(tensors) == 1:
            return tensors[0].contiguous().view(-1)
>       flat = torch.cat([t.contiguous().view(-1) for t in tensors], dim=0)
E       RuntimeError: There were no tensor arguments to this function (e.g., you passed an empty list of Tensors), but no fallback function is registered for schema aten::_cat.  This usually means that this function requires a non-empty list of Tensors.  Available functions are [CPU, CUDA, QuantizedCPU, BackendSelect, Named, AutogradOther, AutogradCPU, AutogradCUDA, AutogradXLA, AutogradNestedTensor, UNKNOWN_TENSOR_TYPE_ID, AutogradPrivateUse1, AutogradPrivateUse2, AutogradPrivateUse3, Tracer, Autocast, Batched, VmapMode].
E
E       CPU: registered at /pytorch/build/aten/src/ATen/RegisterCPU.cpp:5925 [kernel]
E       CUDA: registered at /pytorch/build/aten/src/ATen/RegisterCUDA.cpp:7100 [kernel]
E       QuantizedCPU: registered at /pytorch/build/aten/src/ATen/RegisterQuantizedCPU.cpp:641 [kernel]
E       BackendSelect: fallthrough registered at /pytorch/aten/src/ATen/core/BackendSelectFallbackKernel.cpp:3 [backend fallback]
E       Named: registered at /pytorch/aten/src/ATen/core/NamedRegistrations.cpp:7 [backend fallback]
E       AutogradOther: registered at /pytorch/torch/csrc/autograd/generated/VariableType_2.cpp:9161 [autograd kernel]
E       AutogradCPU: registered at /pytorch/torch/csrc/autograd/generated/VariableType_2.cpp:9161 [autograd kernel]
E       AutogradCUDA: registered at /pytorch/torch/csrc/autograd/generated/VariableType_2.cpp:9161 [autograd kernel]
E       AutogradXLA: registered at /pytorch/torch/csrc/autograd/generated/VariableType_2.cpp:9161 [autograd kernel]
E       AutogradNestedTensor: registered at /pytorch/torch/csrc/autograd/generated/VariableType_2.cpp:9161 [autograd kernel]
E       UNKNOWN_TENSOR_TYPE_ID: registered at /pytorch/torch/csrc/autograd/generated/VariableType_2.cpp:9161 [autograd kernel]
E       AutogradPrivateUse1: registered at /pytorch/torch/csrc/autograd/generated/VariableType_2.cpp:9161 [autograd kernel]
E       AutogradPrivateUse2: registered at /pytorch/torch/csrc/autograd/generated/VariableType_2.cpp:9161 [autograd kernel]
E       AutogradPrivateUse3: registered at /pytorch/torch/csrc/autograd/generated/VariableType_2.cpp:9161 [autograd kernel]
E       Tracer: registered at /pytorch/torch/csrc/autograd/generated/TraceType_2.cpp:10551 [kernel]
E       Autocast: registered at /pytorch/aten/src/ATen/autocast_mode.cpp:254 [kernel]
E       Batched: registered at /pytorch/aten/src/ATen/BatchingRegistrations.cpp:1016 [backend fallback]
E       VmapMode: fallthrough registered at /pytorch/aten/src/ATen/VmapModeRegistrations.cpp:33 [backend fallback]

/home/stas/anaconda3/envs/main-38/lib/python3.8/site-packages/torch/_utils.py:259: RuntimeError
```

**Comment by stas00** [comment]
OK, so I had a look at the first failing test. These 2 can't be separated, since the scheduler needs the optimizer.

But we don't have it yet if it's Deepspeed that creates the optimizer.

So we have a chicken-n-egg problem here.

**Comment by cli99** [comment]
This deepspeed PR https://github.com/microsoft/DeepSpeed/pull/827 fixes the issues. The following tests would pass.
DS scheduler + DS optimizer
HF scheduler + HF optimizer
DS scheduler + HF optimizer 

Shall we put a check in HF to disallow the case HF scheduler + DS optimizer?

**Comment by stas00** [comment]
I tested with your https://github.com/microsoft/DeepSpeed/pull/827 PR tree and indeed 

```
pytest -sv examples/tests/deepspeed/test_deepspeed.py -k test_hf_scheduler_hf_optimizer
pytest -sv examples/tests/deepspeed/test_deepspeed.py -k test_ds_scheduler_hf_optimizer
```

now pass.

> Shall we put a check in HF to disallow the case HF scheduler + DS optimizer?

Correct!

We will need to wait for a new release from your side to commit this PR and add a requirement for that new version. I already have this check setup in another PR waiting for this new release: https://github.com/huggingface/transformers/pull/9624

**Comment by cli99** [comment]
Great. Can you please add the check for the case HF scheduler + DS optimizer? Since you are updating the docs, I think it makes more sense for you to do it. I will work with @jeffra to push the deepspeed PRs into the new release. Thanks.

File: src/transformers/trainer.py
Comment (stas00) [inline src/transformers/trainer.py:620]: @sgugger, if you could validate that it's OK how @cli99 split the optimizer/scheduler creation

File: src/transformers/training_args.py
Comment (sgugger) [inline src/transformers/training_args.py:263]:
```suggestion
        deepspeed (:obj:`str` or :obj:`dict`, `optional`):
```

File: src/transformers/trainer.py
Comment (sgugger) [inline src/transformers/trainer.py:615]:
```suggestion
        Trainer's init through :obj:`optimizers`, or subclass and override this method (or :obj:`create_optimizer` and/or :obj:`create_scheduler`) in a subclass.
```

File: src/transformers/trainer.py
Comment (sgugger) [inline src/transformers/trainer.py:661]:
```suggestion
        Setup the scheduler. The optimizer of the trainer must have been set up before this method is called.
```

Add also the mention of subclassing and overriding if possible.

**Review by sgugger — APPROVED** [review]
Thanks for your PR!

It makes sense to split the optimizer and scheduler creation as you did, and since it's in a BC compatible way I have no worries. Would you mind adding those two new methods in the documentation of the `Trainer` (top of the file main_classes/trainer.rst)?

File: docs/source/main_classes/trainer.rst
Comment (stas00) [inline docs/source/main_classes/trainer.rst:37]: @sgugger - please let me know if this is satisfactory, thank you.

Comment (sgugger) [reply docs/source/main_classes/trainer.rst:37]: Looks good!

**Comment by stas00** [comment]
@cli99, I made further changes to your original code

1. as @jeffra suggested we can't use HF optimizer with offload enabled - so coded to defend against that
2. I realized my original design was flawed and that the user could end up with a mismatch between cl args and the ds config, so I recoded the optimizer/scheduler config sections to override ds config with cl args where needed.

Please let me know if I broke anything in your original plan. I have also updated the docs extensively. They look a bit scary at the moment and will need a rework down the road.

My main goal here is to prevent from user getting subtle errors, so setting command line arguments to override DS config. Hope it makes sense.

**Comment by stas00** [comment]
@sgugger, I made massive doc updates - if you get a chance please kindly skim over them? Thank you!

I think we will merge this on Monday when deepspeed==0.3.13 is planned to be released.

File: docs/source/main_classes/trainer.rst
Comment (sgugger) [inline docs/source/main_classes/trainer.rst:820]:
```suggestion
Here is an example of the pre-configured ``optimizer`` entry for :class:`~transformers.AdamW`:
```

**Review by sgugger — APPROVED** [review]
Just one nit in the documentation. Thanks for all the work!

File: docs/source/main_classes/trainer.rst
Comment (stas00) [reply docs/source/main_classes/trainer.rst:820]: This is DS AdamW :)

I know the docs are a bit hairy at the moment. Will need to take a fresh look once the ZeRO-3 dusts settles.

Comment (stas00) [reply docs/source/main_classes/trainer.rst:820]: Thank you for having another look, @sgugger!

**Comment by sgugger** [comment]
@stas00 I'll let you merge when you are ready (since you followed this more closely than me). It looks good to merge to me :-)
Thanks for your contribution @cli99!

**Comment by stas00** [comment]
I'm on top of this - we are waiting for a new DeepSpeed release required by this PR. Thank you, @sgugger

**Pull request merged**
